### PR TITLE
enable service connection to slave broker via local directory

### PR DESCRIPTION
**Pull request by MartinKT**

### CODE REVIEW

File: libshvbroker/src/rpc/clientconnectiononbroker.cpp [hunk 1]
```diff
@@ -276,6 +282,18 @@ void ClientConnectionOnBroker::onRpcDataReceived(shv::chainpack::Rpc::ProtocolTy
 		}
 		if(m_idleWatchDogTimer)
 			m_idleWatchDogTimer->start();
+		if (isSlaveBrokerConnection() && cp::RpcMessage::isResponse(md)) {
+			int request_id = cp::RpcMessage::requestId(md).toInt();
+			if (m_slaveLsRequestId.contains(request_id)) {
+				cp::RpcValue resp = decodeData(protocol_type, msg_data, 0);
+				shv::chainpack::RpcValue::List result = resp.at(cp::RpcMessage::MetaType::Key::Result).toList();
```
Comment (fvacek) [inline libshvbroker/src/rpc/clientconnectiononbroker.cpp:289]:
```c++
cp::RpcResponse resp2(resp);
resp2.setResult(...);
```

Comment (fvacek) [inline libshvbroker/src/rpc/clientconnectiononbroker.cpp:289]: handle error

File: libshvbroker/src/rpc/clientconnectiononbroker.cpp [hunk 2]
```diff
@@ -276,6 +282,31 @@ void ClientConnectionOnBroker::onRpcDataReceived(shv::chainpack::Rpc::ProtocolTy
...
+			int request_id = cp::RpcMessage::requestId(md).toInt();
+			auto it = std::find_if(m_slaveLsRequestId.begin(), m_slaveLsRequestId.end(), [&request_id](const RequestId &r) {
+				return r.requestId == request_id;
+			});
+			if (it != m_slaveLsRequestId.end()) {
+				QVector<int> caller_ids = makeCallerIdList(cp::RpcMessage::callerIds(md));
+				if (caller_ids == it->callerIds) {
+					m_slaveLsRequestId.erase(it);
+					cp::RpcValue resp = decodeData(protocol_type, msg_data, 0);
+					if (!resp.at(cp::RpcMessage::MetaType::Key::Error).isValid()) {
```
Comment (fvacek) [inline libshvbroker/src/rpc/clientconnectiononbroker.cpp:295]: what `if (resp.at(cp::RpcMessage::MetaType::Key::Error).isValid()) ` ?

File: libshvbroker/src/rpc/clientconnectiononbroker.h [hunk 3]
```diff
@@ -57,11 +57,19 @@ class ClientConnectionOnBroker : public shv::iotqt::rpc::ServerConnection, publi
 	void onSocketConnectedChanged(bool is_connected);
 	void onRpcDataReceived(shv::chainpack::Rpc::ProtocolType protocol_type, shv::chainpack::RpcValue::MetaData &&md, std::string &&msg_data) override;
 	bool checkTunnelSecret(const std::string &s);
+	QVector<int> makeCallerIdList(const shv::chainpack::RpcValue &caller_ids);
```
Comment (fvacek) [inline libshvbroker/src/rpc/clientconnectiononbroker.h:60]: `static QVector<int> callerIdsToList(const shv::chainpack::RpcValue &caller_ids);`

File: libshvbroker/src/rpc/clientconnectiononbroker.h [hunk 4]
```diff
@@ -57,11 +57,19 @@ class ClientConnectionOnBroker : public shv::iotqt::rpc::ServerConnection, publi
...
 
 	void processLoginPhase() override;
 private:
 	QTimer *m_idleWatchDogTimer = nullptr;
 	std::string m_mountPoint;
+
+	struct RequestId
```
Comment (fvacek) [inline libshvbroker/src/rpc/clientconnectiononbroker.h:67]: `RequestAndCallersIds`

File: libshvbroker/src/rpc/masterbrokerconnection.cpp [hunk 5]
```diff
@@ -138,7 +138,14 @@ void MasterBrokerConnection::onRpcDataReceived(shv::chainpack::Rpc::ProtocolType
 			return;
 		}
 		if(cp::RpcMessage::isRequest(md)) {
-			shv::core::String shv_path = masterExportedToLocalPath(cp::RpcMessage::shvPath(md).toString());
+			shv::core::String request_shv_path = cp::RpcMessage::shvPath(md).toString();
```
Comment (fvacek) [inline libshvbroker/src/rpc/masterbrokerconnection.cpp:141]: `const shv::core::String &request_shv_path = cp::RpcMessage::shvPath(md).asString();`

**Review by fvacek — CHANGES_REQUESTED** [review]

File: libshvbroker/src/rpc/clientconnectiononbroker.cpp (hunk 2 above)
Comment (MartinKT) [reply libshvbroker/src/rpc/clientconnectiononbroker.cpp:295]: Nothing, reponse is not modified and is proceeded vid default handler.

File: libshvbroker/src/rpc/clientconnectiononbroker.h (hunk 4 above)
Comment (MartinKT) [reply libshvbroker/src/rpc/clientconnectiononbroker.h]: fixed

File: libshvbroker/src/rpc/clientconnectiononbroker.h (hunk 3 above)
Comment (MartinKT) [reply libshvbroker/src/rpc/clientconnectiononbroker.h]: fixed

**Pull request merged**